File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:

    int preOrder(TreeNode* root) {
        if (!root) return 0;
        int left = preOrder(root->left);
        int right = preOrder(root->right);
        return 1 + max(left, right);
    }

    int maxDepth(TreeNode* root) {
        if (!root) return 0;
        return preOrder(root);
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] right = new int[n+1];
        right[n] = 1;
        for (int i=n-1; i>=0; i--) {
            right[i] = right[i+1] * nums[i];
        }
        int left = 1;
        int[] ans = new int[n];
        for (int i=0; i<n; ++i) {
            ans[i] = left * right[i+1];
            left *= nums[i];
        }
        return ans;
    }
}  
